core/src/repair/quic_endpoint.rs

```diff
@@ -444,195 +567,77 @@ async fn handle_connection(
     }
 }
 
-async fn recv_requests_task(
+async fn read_datagram_task<T>(
     endpoint: Endpoint,
     remote_address: SocketAddr,
     remote_pubkey: Pubkey,
     connection: Connection,
-    remote_request_sender: Sender<RemoteRequest>,
-    stats: Arc<RepairQuicStats>,
-) -> Result<(), Error> {
-    loop {
-        let (send_stream, recv_stream) = connection.accept_bi().await?;
-        tokio::task::spawn(handle_streams_task(
-            endpoint.clone(),
-            remote_address,
-            remote_pubkey,
-            send_stream,
-            recv_stream,
-            remote_request_sender.clone(),
-            stats.clone(),
-        ));
-    }
-}
-
-async fn handle_streams_task(
-    endpoint: Endpoint,
-    remote_address: SocketAddr,
-    remote_pubkey: Pubkey,
-    send_stream: SendStream,
-    recv_stream: RecvStream,
-    remote_request_sender: Sender<RemoteRequest>,
+    sender: Sender<T>,
     stats: Arc<RepairQuicStats>,
-) {
-    if let Err(err) = handle_streams(
-        &endpoint,
-        remote_address,
-        remote_pubkey,
-        send_stream,
-        recv_stream,
-        &remote_request_sender,
-    )
-    .await
-    {
-        debug!("handle_stream: {remote_address}, {remote_pubkey}, {err:?}");
-        record_error(&err, &stats);
-    }
-}
-
-async fn handle_streams(
-    endpoint: &Endpoint,
-    remote_address: SocketAddr,
-    remote_pubkey: Pubkey,
-    mut send_stream: SendStream,
-    mut recv_stream: RecvStream,
-    remote_request_sender: &Sender<RemoteRequest>,
-) -> Result<(), Error> {
+) -> Result<(), Error>
+where
+    T: From<(Pubkey, SocketAddr, Bytes)>,
+{
     // Assert that send won't block.
-    debug_assert_eq!(remote_request_sender.capacity(), None);
-    const READ_TIMEOUT_DURATION: Duration = Duration::from_secs(2);
-    let bytes = tokio::time::timeout(
-        READ_TIMEOUT_DURATION,
-        recv_stream.read_to_end(PACKET_DATA_SIZE),
-    )
-    .await
-    .map_err(|_| Error::ReadToEndTimeout)??;
-    let (response_sender, response_receiver) = tokio::sync::oneshot::channel();
-    let remote_request = RemoteRequest {
-        remote_pubkey: Some(remote_pubkey),
-        remote_address,
-        bytes,
-        response_sender: Some(response_sender),
-    };
-    if let Err(err) = remote_request_sender.send(remote_request) {
-        close_quic_endpoint(endpoint);
-        return Err(Error::from(err));
-    }
-    let Ok(response) = response_receiver.await else {
-        return Err(Error::NoResponseReceived);
-    };
-    for chunk in response {
-        let size = chunk.len() as u64;
-        send_stream.write_all(&size.to_le_bytes()).await?;
-        send_stream.write_all(&chunk).await?;
+    debug_assert_eq!(sender.capacity(), None);
+    loop {
+        match connection.read_datagram().await {
+            Ok(bytes) => {
+                let value = T::from((remote_pubkey, remote_address, bytes));
+                if let Err(err) = sender.send(value) {
```

Comment:
please add an error log

Reply:
How do we guard using excessive amount of memory here?

Reply:
> please add an error log

This is returning the error to the caller which then is recorded into the metrics.
An error log would probably be too chatty here.

> How do we guard using excessive amount of memory here?

The other end of the channel does load shedding for repair requests.
Repair responses also should come back with correct nonce and otherwise are dropped.

Reply:
I am still concerned what if the other end is so busy it could not shed the requests fast enough? I think from defensive programming point, the server need to bound the queue to a reasonable size. @behzadnouri

Reply:
okay, I will look into using bounded channels instead.